src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -1032,6 +1021,24 @@ private static object ReplaceOperatorImpl(ExecutionContext context, string input
                     string replacement = PSObject.ToStringParser(context, substitute);
                     return regex.Replace(input, replacement);
             }
+
+            // Local helper function to avoid creating an instance of the generated delegate helper class
+            // every time 'ReplaceOperatorImpl' is invoked.
+            static MatchEvaluator GetMatchEvaluator(ExecutionContext context, ScriptBlock sb)
```

Comment:
We could put the method in a helper static class and save the result in an private static. So we will create the delegate only once.
I found the pattern in https://www.stevejgordon.co.uk/high-performance-logging-in-net-core.

Reply:
Does the compiler not cache the delegate?

Reply:
@iSazonov The delegate body for `MatchEvaluator` requires 2 closed-over variables from its lexically-enclosing method `sb` and `context` -- they are not passed to the delegate as paramters, and thus a generated helper class is not avoidable.

The article you referred to is a good read, but the key perf improvement there is avoid processing the message template for every call to each of the log methods.
> The key improvement here, vs. Standard logging, is that the template for the log message now only needs to be parsed once when this message is defined. In standard logging, that occurs for each call.

Reply:
@xtqqczze The delegate here cannot be cached becuase it depends on the scirpt block `sb` and also the execution context `context`. These 2 closed-over variables make caching not useful in most cases.

Reply:
> The delegate body for MatchEvaluator requires 2 closed-over variables from its lexically-enclosing method sb and context -- they are not passed to the delegate as paramters, and thus a generated helper class is needed.

Can't we pass them as parameters?
I guess the pattern I mentioned excludes delegate allocations at all.

Reply:
No we cannot. The delegate `MatchEvaluator` only accepts a `Match` type parameter: https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.matchevaluator?view=net-5.0

> I guess the pattern I mentioned excludes delegate allocations at all.

The article you refered to is not for reducing delegate allocation. The "STANDARD LOGGING" it compares to doesn't involve any delegation creation.

Reply:
We seem to be talking about different things. :-)
One code path where the ReplaceOperatorImpl() is called is a `while` cycle. There the same context and the same scriptblock but we create the delegate again and again - what if the lval is a large list?

Reply:
I see what you were talking about now :)
I'm not trying to optimize for all scenarios, but the most common scenario.

To optimize for the scenario you are talking about, you will need a delegate cache with both the scirpt block and the context together as the key. But as I said previously, the cache will not be useful for most cases. You need to think about how common would `$ABigList -replace "pattern", $AScriptBlock` be used. Holding the cache indefinitely would be a waste of memory; adding logic to clear the cache smartly would be extra complexity. The value would not be worth the effort.

Reply:
> Either way, I don't think the value would be worth the effort.

After bad experience, I avoid such scripts. I suppose others do the same. But this does not mean that it cannot be improved. For example, the original script from the question could be rewritten using this pattern and if optimized, it could possibly run even faster than the original script.

I will open a tracking issue for the optimization.

Reply:
> For example, the original script from the question could be rewritten using this pattern and if optimized ...

No, the original repro scirpt cannot be rewritten with `$ABigList -replace "pattern", $AScriptBlock`, because
1. the script uses many different patterns
2. the substitute strings are known, not requiring calculation based on `$match`. (be noted on the `$AScriptBlock` part)

Reply:
I am definitely in another dimension :-)